### PR TITLE
Htt 9 4 11 cand1 v1

### PR DESCRIPTION
**Pull request by GaelTouquet**
updates:
- MET v2
- Embedding
- minor sync updates (JEC, TES propagation, TauID called pre-TES)
- crab submitting minor updates
- got rid of some old sizeable files to be able to submit crab jobs using the latest JEC files

### CODE REVIEW

File: H2TauTau/python/heppy/analyzers/JetAnalyzer.py [hunk 1]
```diff
@@ -35,15 +36,25 @@ def declareHandles(self):
         )
         
     def process(self, event):
+
+        pt_cut = 50.0
```
Comment (cbernet) [inline H2TauTau/python/heppy/analyzers/JetAnalyzer.py]: in the configuration of the analyzer, the user should provide a lambda
```
is_badjet = lambda ....
```

File: H2TauTau/python/heppy/analyzers/JetAnalyzer.py [hunk 2]
```diff
@@ -35,15 +36,25 @@ def declareHandles(self):
...
+        eta_min = 2.65
+        eta_max = 3.139
+
         self.readCollections(event.input)
         jets = self.handles['jets'].product()
         output_jets = []
+        output_jets_to_correct = []
         for jet in jets:
             hjet = Jet(jet)
+            if ( hjet.correctedJet("Uncorrected").pt() > pt_cut or abs(hjet.eta()) < eta_min or abs(hjet.eta()) > eta_max ) :
```
Comment (cbernet) [inline H2TauTau/python/heppy/analyzers/JetAnalyzer.py]: use the lambda directly

File: H2TauTau/python/heppy/analyzers/METAnalyzer.py [hunk 3]
```diff
@@ -8,12 +8,23 @@
...
 from ROOT import HTTRecoilCorrector as RC
 
 LorentzVector = ROOT.Math.LorentzVector(ROOT.Math.PxPyPzE4D("double"))
 
+def get_final_ptcs(ptc):
+    if ptc.numberOfDaughters() == 0 :
+        return [ptc]
+    else :
+        final_ptcs = []
+        for N in range(ptc.numberOfDaughters()):
```
Comment (cbernet) [inline H2TauTau/python/heppy/analyzers/METAnalyzer.py]: avoid single letter variables, except for i, j in enumeration. 
don't use single capital variables. 
here ndau would be good

File: H2TauTau/python/heppy/analyzers/METAnalyzer.py [hunk 4]
```diff
@@ -94,34 +115,48 @@ def process(self, event):
...
 
+        # Correct PF MET
+        pfmet_px_old = event.pfmet.px()
+        pfmet_py_old = event.pfmet.py()
+
+        if hasattr(self.cfg_ana, 'runFixEE2017') and self.cfg_ana.runFixEE2017:
+            rawMET = self.runFixEE2017(event)
+            pfmet_px_old = rawMET.px()
+            pfmet_py_old = rawMET.py()
+
```
Comment (cbernet) [inline H2TauTau/python/heppy/analyzers/METAnalyzer.py]: rm line

File: H2TauTau/python/heppy/analyzers/METAnalyzer.py [hunk 5]
```diff
@@ -94,34 +115,48 @@ def process(self, event):
...
+        if hasattr(self.cfg_ana, 'runFixEE2017') and self.cfg_ana.runFixEE2017:
+            rawMET = self.runFixEE2017(event)
+            pfmet_px_old = rawMET.px()
+            pfmet_py_old = rawMET.py()
+            
+            if event.type1METCorr :
+                pfmet_px_old += event.type1METCorr[0]
+                pfmet_py_old += event.type1METCorr[1]
+
```
Comment (cbernet) [inline H2TauTau/python/heppy/analyzers/METAnalyzer.py]: rm line

File: H2TauTau/python/heppy/analyzers/NtupleProducer.py [hunk 6]
```diff
@@ -17,4 +17,7 @@ def process(self, event):
                 except OverflowError:
                     print 'value', var.function(data), "didn't fit in var", varname
                     continue
+                except TypeError:
+                    print varname, 'took the wrong type'
```
Comment (cbernet) [inline H2TauTau/python/heppy/analyzers/NtupleProducer.py]: clarify error message

File: H2TauTau/python/heppy/analyzers/TauP4Scaler.py [hunk 7]
```diff
@@ -37,6 +37,8 @@ def correct_energy(self, tau):
                 if decayMode in self.decay_modes_dict.keys():
                     energy_scale = TauEnergyScales[ self.gen_match_dict[gen_match] ][ self.decay_modes_dict[decayMode] ]
             tau.unscaledP4 = copy.copy(tau.p4())
+            tau.tauID('byIsolationMVArun2017v2DBoldDMwLTraw2017')
```
Comment (cbernet) [inline H2TauTau/python/heppy/analyzers/TauP4Scaler.py:41]: add a comment before these two lines
explain in the gitlab

File: H2TauTau/python/heppy/ntuple/ntuple_variables.py [hunk 8]
```diff
@@ -9,7 +9,7 @@
     lumi = v(lambda x: x.lumi, int),
     event = v(lambda x: x.eventId, int, 'l'),
     n_up = v(lambda x: getattr(x, 'NUP', default), int),
-    n_pu = v(lambda x: x.nPU if x.nPU is not None else default),
+    n_pu = v(lambda x: getattr(x, 'nPU', default) if getattr(x, 'nPU', default) is not None else default, int),
```
Comment (cbernet) [inline H2TauTau/python/heppy/ntuple/ntuple_variables.py]: add comments for all these changes (at least write that you cover specificities of the embedded samples )

File: H2TauTau/python/heppy/ntuple/ntuple_variables.py [hunk 9]
```diff
@@ -37,7 +37,7 @@
 ]
 event_flags = Block('event_flags', lambda x: x)
 for flag in flags: 
-    event_flags[flag] = v(lambda x: getattr(x,flag), int)
+    event_flags[flag] = v(lambda x: getattr(x,flag,1), int)
```
Comment (cbernet) [inline H2TauTau/python/heppy/ntuple/ntuple_variables.py]: and here explain why chosing 1 (no flag is good )

File: H2TauTau/python/heppy/sequence/common.py [hunk 10]
```diff
@@ -236,7 +236,7 @@ def select_electron_third_lepton_veto(electron):
     'jets_20_unclean',
     output = 'jets_20_unclean',
     src = 'jets',
-    filter_func = lambda x : x.pt()>20 and abs(x.eta())<4.7 and x.jetID("POG_PFID_Tight")
+    filter_func = lambda x : x.pt()>20 and abs(x.eta())<4.7 and x.jetID("POG_PFID_Tight") and not ( x.pt() < 50 and abs(x.eta()) > 2.65 and abs(x.eta()) < 3.139 )
```
Comment (cbernet) [inline H2TauTau/python/heppy/sequence/common.py]: line too long, break it \

**Review by cbernet — CHANGES_REQUESTED** [review]

**Comment by GaelTouquet** [comment]
requested changes have been added

File: H2TauTau/python/heppy/analyzers/NtupleProducer.py [hunk 11]
```diff
@@ -18,6 +18,11 @@ def process(self, event):
                     print 'value', var.function(data), "didn't fit in var", varname
                     continue
                 except TypeError:
-                    print varname, 'took the wrong type'
-                    raise
+                    message = ' '.join(['Variable',
```
Comment (cbernet) [inline H2TauTau/python/heppy/analyzers/NtupleProducer.py]: please use format

Comment (GaelTouquet) [reply H2TauTau/python/heppy/analyzers/NtupleProducer.py]: Done

**Pull request merged**